laziness tutorial

joelgrus opened the pull request:
this ended up being kind of tricky to write, let me know if you can think of ways to improve it

matt-gardner commented on tutorials/getting_started/laziness.md:
```diff
@@ -0,0 +1,233 @@
+to `MyDatasetReader._read()`, and your instances will be read from disk
+10 times.
+
+In this case, you should make sure that your `_read()` method
```
I'm not sure what the right way is to talk about this, but if you're sampling things at each call to `_read()`, you actually _do_ want to read the whole dataset into memory, but only on the _first_ call to `_read()`.  Then, after reading everything that you want, all calls to `_read()` do the sampling on the instances in memory, and it's different at each epoch.  This is an advanced usage that you might not want to cover here, but I'd at least avoid contradicting this use case in the simple tutorial.

joelgrus replied:
(1) why would you implement sampling in the DatasetReader rather than as a `SamplingIterator` or something?
(2) doesn't this depend on how you sample? you could take each instance with a fixed probability without loading the whole dataset, or you could do some kind of reservoir sampling.
(3) all that said, I guess I can imagine cases where you would want to sample in the DatasetReader, so let me think about how to word that

matt-gardner replied:
The case I'm thinking of is that you have `k`paragraphs associated with each instance, and at each epoch you want to sample 2 of them. You read in all `k`, cache that, then pick two at each call to`_read`. You can't easily do this kind of thing in our `DataIterator` - you're already dealing with`Instances`by that point.

And by the way, this is how I recommend doing a `DatasetReader` for Chris's model.

matt-gardner commented on tutorials/getting_started/laziness.md:
```diff
@@ -0,0 +1,233 @@
+The other option here is `instances_per_epoch`. By default, each epoch is a single
+pass through the dataset. However, if you had millions of instances,
+you might want each epoch to consist of a fraction of the dataset,
+in which case you can specify the value here. But most likely you don't need
```
Specifying this will also result in nicer tqdm logging.

joelgrus replied:
the way things are designed right now that's a problem unless your train / validation / test sets are the same size.

i.e. if you have 1000 train examples / 200 validation examples, and if you specify `instances_per_epoch: 1000`, then when you iterate over the validation set it will make 5 passes over the data

this is why I was earlier suggesting making the dataset parts of the configuration slightly richer -- it seems likely you'd want to specify different values for things like this for different datasets

joelgrus replied:
or another alternative is just to count the instances the first pass through the dataset and remember the number for future passes

matt-gardner replied:
Oh, yeah, you're right, I forgot about that.  Mandar said he was going to submit a PR to allow for different train and validation iterators, which would solve this problem.

DeNeutoy commented on tutorials/getting_started/laziness.md:
````diff
@@ -0,0 +1,233 @@
+lazy iterables (again, I stripped it down to just its essence):
+
+```python
+class _LazyInstances(Iterable):
````
This is pretty detailed and involves a private AllenNLP class. I'm not sure the details really belong here?

DeNeutoy replied:
* This level of detail

matt-gardner replied:
On the other hand, this is a tutorial, not a how-to, and the goal is to teach the reader about how the library works.  We _definitely_ want to mention in here that you'll get multiple calls to `_read()` with `lazy=True`; I could go either way on whether to include this code to demonstrate that.

matt-gardner commented:
Also, I just realized we need to update the README for the tutorials, that has an index of them.

joelgrus commented:
ok, I think I addressed all the concerns (except for Mark's concerns about talking about the details of a private class, I didn't address those 😇 ), I also updated the README

matt-gardner commented on tutorials/how_to/laziness.md:
```diff
@@ -0,0 +1,31 @@
+ possibly because you want to start training immediately,
+ possibly for some other reason I haven't thought of.)
+
+There are basically four things you should do. This tutorial will tell you
```
`s/tutorial/how-to/`?

matt-gardner commented on tutorials/how_to/laziness.md:
```diff
@@ -0,0 +1,31 @@
+
+## In `YourDatasetReader._read()`
+
+`YourDatasetReader._read()` should return a generator, not a list.
```
Is it possible to link to particular code blocks in the other tutorial for `generator` and `list`, in case someone doesn't know what this means?  Worst case, you could just link to the same thing for `generator` that you did in the tutorial.

matt-gardner commented on tutorials/how_to/laziness.md:
```diff
@@ -0,0 +1,31 @@
+
+You should specify `'max_instances_in_memory'` as large as you can
+in the `iterator` section of your experiment config.
+(You don't _have_ to do this, but it's a good idea.)
```
I'd add "particularly if you want to use a bucket iterator".
